pyrtls.pyi

```diff
@@ -0,0 +1,407 @@
+        """
+        How many bytes could be written by `Connection.write_tls_into()` if called right now.
+        A non-zero value implies that `Connection.wants_write()` would yield `True`.
+        """
+
+    def plaintext_bytes_to_read(self) -> int:
+        """How many plaintext bytes are currently buffered in the connection."""
+
+    def peer_has_closed(self) -> bool:
+        """
+        `True` if the peer has sent us a `close_notify` alert. This is the TLS mechanism to
+        securely half-close a TLS connection, and signifies that the peer will not send any
+        further data on this connection.
+        """
+
+class TLSError: ...
+
+# client.rs
+
+class ClientSocket:
+    """
+    A `ClientSocket` is a wrapper type that contains both a `socket.socket` and a
+    `ClientConnection` object. It is similar to the `ssl.SSLSocket` class from the
+    standard library and should implement most of the same methods.
+    """
+
+    def connect(self, address: tuple[str, int]) -> None:
+        """
+        Connect to a remote socket address. `address` must currently be a 2-element
+        tuple containing a hostname and a port number.
+        """
+
+    def do_handshake(self) -> None:
+        """Perform the TLS setup handshake."""
+
+    def send(self, bytes: bytes) -> int:
+        """
+        Send data to the socket. The socket must be connected to a remote socket. Returns the
+        number of bytes sent. Applications are responsible for checking that all data has been
+        sent; if only some of the data was transmitted, the application needs to attempt delivery
+        of the remaining data.
+        """
+
+    def recv(self, size: int) -> bytes:
+        """
+        Receive data from the socket. The return value is a bytes object representing the data
+        received. The maximum amount of data to be received at once is specified by `size`.
+        A returned empty bytes object indicates that the server has disconnected.
+        """
+
+class ClientConnection:
+    """
+    A `ClientConnection` contains TLS state associated with a single client-side connection.
+    It does not contain any networking state, and is not directly associated with a socket,
+    so I/O happens via the methods on this object directly.
+
+    A `ClientConnection` can be created from a `ClientConfig` `config` and a server name, `name`.
+    The server name must be either a DNS hostname or an IP address (only string forms are
+    currently accepted).
+    """
+
+    def __new__(cls, config: ClientConfig, name: str) -> ClientConnection: ...
+    def readable(self) -> bool:
+        """
+        Returns `true` if the caller should call `read_tls()` as soon as possible.
+
+        If there is pending plaintext data to read, this returns `false`. If your application
+        respects this mechanism, only one full TLS message will be buffered by pyrtls.
+        """
+
+    def read_tls(self, buf: list[int]) -> int:
```

Comment:
In the `ClientConnection.read_tls` and similar methods, the Rust implementation takes a `&[u8]`. I put `list[int]` as the Python version of this, but I'm not sure if this is right. Maybe it should be a `tuple[int]`? Maybe this method shouldn't exist at all in the stub?

Reply:
If I understand the PyO3 documentation correctly, it uses the [`FromPyObject`](https://docs.rs/pyo3/latest/pyo3/conversion/trait.FromPyObject.html) trait to convert, and this requires a `bytes` object where we take a `&[u8]`.

Reply:
That makes sense. Fixed in the next commit.